### ISSUE TITLE
Wrong behavior with onlyOn and onlyForStage options.

### ISSUE THREAD

**Issue by karakum** [deployphp/deployer#1038]
| Q                 | A
| ----------------- | ---
| Issue Type        | Bug
| Deployer Version  | 4.1.0
| Local Machine OS  | linux
| Remote Machine OS | linux

### Description 
Wrong behavior with `onlyOn` and `onlyForStage` options.

### Content of `deploy.php`
```php
<?php
namespace Deployer;
require 'recipe/common.php';


localServer('s1')
    ->stage('develop');

localServer('s2')
    ->stage('production');


task('task:bad', function () {
    if (in_array('develop', get('stages'))) {
        writeln('task on develop');
    } elseif (in_array('production', get('stages'))) {
        writeln('task on production');
    }
})->desc('Task bad solution');

desc('Run task for s1 server');
task('task:s1', function () {
    writeln('task only s1');
})->onlyOn('s1');

desc('Run task for s2 server');
task('task:s2', function () {
    writeln('task only s2');
})->onlyOn('s2');

desc('Run task for develop stage');
task('task:develop', function () {
    writeln('task only develop');
})->onlyForStage('develop');

desc('Run task for production stage');
task('task:production', function () {
    writeln('task only production');
})->onlyForStage('production');

task('test', [
    'task:bad',
    'task:s1',
    'task:s2',
    'task:develop',
    'task:production',
])->desc('Test condition executing');

```

### Output log
1. Run by server s1(develop)
```
$ dep -vvv test s1
➤ Executing task task:bad
task on develop
• done on [s1]
✔ Ok [0ms]
➤ Executing task task:s1
task only s1
• done on [s1]
✔ Ok [0ms]
➤ Executing task task:s2
✔ Ok [0ms]
```
**Expected `Executing task task:develop`**

2. Run by server s2(production)
```
$ dep -vvv test s2
➤ Executing task task:bad
task on production
• done on [s2]
✔ Ok [0ms]
➤ Executing task task:s1
✔ Ok [0ms]
➤ Executing task task:s2
task only s2
• done on [s2]
✔ Ok [0ms]
```
**Expected `Executing task task:production`**

3. Run by stage develop(s1)
```
$ dep -vvv test develop
➤ Executing task task:bad
task on develop
• done on [s1]
✔ Ok [0ms]
➤ Executing task task:s1
task only s1
• done on [s1]
✔ Ok [0ms]
➤ Executing task task:s2
✔ Ok [0ms]
➤ Executing task task:develop
task only develop
• done on [s1]
✔ Ok [0ms]
```
**Work as expected.**

4. Run by stage production(s2)
```
$ dep -vvv test production
➤ Executing task task:bad
task on production
• done on [s2]
✔ Ok [0ms]
➤ Executing task task:s1
✔ Ok [0ms]
➤ Executing task task:s2
task only s2
• done on [s2]
✔ Ok [0ms]
➤ Executing task task:production
task only production
• done on [s2]
✔ Ok [1ms]
```
**Work as expected.**

In all four cases there are indication `Executing task task:s1` and `Executing task task:s2` but actual output seems correct. 
I suppose that:
- first case should contain `Executing task task:develop`
- second case should contain `Executing task task:production`
- first and third cases should not contain `Executing task task:s1`
- second and fourth cases should not contain `Executing task task:s2`

Am I wrong?

**Comment by elfet** [comment]
Hi, i'm thinkng about refactoring the whole system in v5. Now i'm thinking about better api.

**Comment by karakum** [comment]
Isn't this is BUG ? Why feature?
Task 'onlyOn' does not executed if deploy call by stage name

**Comment by elfet** [comment]
You're right. It's a bug.

**Comment by antonmedv** [comment]
Fixed in #1092.
